internal/replication/blockreplicator.go

```diff
@@ -638,6 +650,21 @@ func (br *BlockReplicator) updateLastProposal(lastBlockProposed *types.Block) {
 		} else {
 			br.lg.Panicf("Failed to compute last block base hash: %s", err)
 		}
+		br.numInFlightBlocks++
+
+		if br.numInFlightBlocks > br.clusterConfig.ConsensusConfig.RaftConfig.MaxInflightBlocks {
+			br.lg.Infof("Number of in-flight blocks %d exceeds max %d, waiting for blocks to commit",
+				br.numInFlightBlocks, br.clusterConfig.ConsensusConfig.RaftConfig.MaxInflightBlocks)
+
+			for br.numInFlightBlocks > br.clusterConfig.ConsensusConfig.RaftConfig.MaxInflightBlocks {
```

Comment:
I think the `numInFlightBlocks` need to be an atomic variable so that we can load and set it atomically.

Reply:
This is in a method 
```
// updateLastProposal updates the last block proposed in order to keep track of block numbering.
func (br *BlockReplicator) updateLastProposal(lastBlockProposed *types.Block) {
	br.mutex.Lock()
	defer br.mutex.Unlock()
```

In general, `numInFlightBlocks` must be changed with a mutex locked for the waiting go-routine (event-loop) to see the change reliably after it wakes up when Broadcast() is called. In addition, for the two routines to communicate reliably, a mutex must be used, as several variables are changed by the two routines, e.g. numInFlightBlocks, lastCommittedBlock, etc. An atomic is not good for that as it does not have the "happen before" guaranties for multiple variables at once.